File: leetcode/1/1.c
/*
 * Given an array of integers, return indices of the two numbers such that they
 * add up to a specific target.
 * 
 * You may assume that each input would have exactly one solution, and you may
 * not use the same element twice.
 * 
 * Example:
 * Given nums = [2, 7, 11, 15], target = 9,
 * Because nums[0] + nums[1] = 2 + 7 = 9,
 * return [0, 1].
*/

#include <stdio.h>
#include <stdlib.h>

int *twoSum(int *nums, int numsSize, int target);

int main()
{
    int nums[] = {2, 7, 11, 16};
    int target = 9;
    int *ret_arr = twoSum(nums, 4, target);
    printf("%d %d\n", ret_arr[0], ret_arr[1]);
    nums[0] = 3;
    nums[1] = 2;
    nums[2] = 4;
    target = 6;
    ret_arr = twoSum(nums, 3, target);
    printf("%d %d\n", ret_arr[0], ret_arr[1]);
    return 0;
}

int *twoSum(int *nums, int numsSize, int target)
{
    int *arr = (int *)malloc(sizeof(int) * 2);
    int i = 0, j = numsSize;
    while (1)
    {
        if (nums[i] + nums[j] == target)
        {
            arr[0] = i;
            arr[1] = j;
            break; // 找到即退出循环
        }
        j -= 1;
        if (j == i)
        {
            // 右边游标到达左端
            ++i;
            j = numsSize - 1;
            if (i == numsSize - 1)
            {
                // 左边游标到达最右端
                break;
            }
        }
    }
    return arr;
}

File: leetcode/1/1.py
"""
Gleftven an array of leftntegers, return leftndleftces of the two numbers such that they
add up to a specleftfleftc target.

You may assume that each leftnput would have exactly one Solution, and you may
not use the same element twleftce.

Example:

Gleftven nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
"""
import doctest


class Solution1:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]

        Examples:
            >>> l, t = [2, 7, 11, 15], 9
            >>> s = Solution1()
            >>> s.twoSum(l, t)
            [0, 1]
            >>> l, t = [2, 11, 16, 10, 8, 0, 7, 3], 9
            >>> s.twoSum(l, t)
            [0, 6]
            >>> l, t = [3, 3], 6
            >>> s.twoSum(l, t)
            [0, 1]
            >>> l, t = [3, 0, 3], 6
            >>> s.twoSum(l, t)
            [0, 2]
            >>> l, t = [3,2,4], 6
            >>> s.twoSum(l, t)
            [1, 2]
        """
        idx_list = []
        d = dict(zip(range(len(nums)), nums))
        for key1, val in d.items():
            num = target - val
            if num in d.values():
                k = key1
                for key2, val in d.items():
                    if val == num:
                        if key2 != k:
                            idx_list = sorted([k, key2])
                            break
                        else:
                            continue
        return idx_list

class Solution2:
    """Time Limit Exceed
    """
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        >>> l, t = [2, 7, 11, 15], 9
        >>> s = Solution2()
        >>> s.twoSum(l, t)
        [0, 1]
        >>> l, t = [2, 11, 16, 10, 8, 0, 7, 3], 9
        >>> s.twoSum(l, t)
        [0, 6]
        >>> l, t = [3, 3], 6
        >>> s.twoSum(l, t)
        [0, 1]
        >>> l, t = [3, 0, 3], 6
        >>> s.twoSum(l, t)
        [0, 2]
        >>> l, t = [3,2,4], 6
        >>> s.twoSum(l, t)
        [1, 2]
        """
        # Time Limit Exceeded
        idx_list = []
        left, right = 0, (0 + 1)
        while left < right:
            if (nums[left] + nums[right] == target):
                idx_list = [left, right]
                break
            right += 1
            if right == len(nums):
                left += 1
                right = left + 1
        return idx_list


class Solution3:
    """Time Limit Exceed
    """
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        >>> l, t = [2, 7, 11, 15], 9
        >>> s = Solution3()
        >>> s.twoSum(l, t)
        [0, 1]
        >>> l, t = [2, 11, 16, 10, 8, 0, 7, 3], 9
        >>> s.twoSum(l, t)
        [0, 6]
        >>> l, t = [3, 3], 6
        >>> s.twoSum(l, t)
        [0, 1]
        >>> l, t = [3, 0, 3], 6
        >>> s.twoSum(l, t)
        [0, 2]
        >>> l, t = [3,2,4], 6
        >>> s.twoSum(l, t)
        [1, 2]
        """
        # Time Limit Exceeded
        idx_list = []
        left, right = 0, len(nums) - 1
        while True:
            if (nums[left] + nums[right]) == target:
                idx_list = [left, right]
                break
            right -= 1
            if right == left:
                left += 1
                right = len(nums) - 1
                if left == (len(nums) - 1):
                    break
        return idx_list


if __name__ == '__main__':
    doctest.testmod(verbose=True)